### PR TITLE
Refactor FXIOS-8188 [v124] [Multi-window] Redux refactors (Tab, BVC, Toasts) for multi-window

### PR DESCRIPTION
**Pull request by mattreaganmozilla**
## :scroll: Tickets
[Jira ticket](https://mozilla-hub.atlassian.net/browse/FXIOS-8188)
[Github issue](https://github.com/mozilla-mobile/firefox-ios/issues/18209)

## :bulb: Description

Part of ongoing efforts to update our Redux stack to work with multiple iPad windows. I tried to keep the PR as small as possible but these changes impact an entire chain of logic in our Redux engine that requires a number of updates in order to function properly. Several of the remaining Redux components (**ThemeManager**, **FakeSpot**, **Felt Privacy** etc.) have been left unchanged; those will be updated in a separate PR.

Summary of changes:
- Begins moving our `Action` enums towards using a consistent pattern of including an `ActionContext` (or concrete subclass) which includes the window UUID
- Updates `windowUUID` to be non-optional (some code is using `.unavailable` for now, this is largely temporary)
- Refactors to related middleware, reducers, etc. for these changes

### Still forthcoming

- Updates to unit tests will be pushed up once the PR is reviewed, in order to minimize noise in the diff
- Additional refactors to other Redux/multi-window components will be forthcoming in follow-up PRs

## :pencil: Checklist
You have to check all boxes before merging
- [x] Filled in the above information (tickets numbers and description of your work)
- [x] Updated the PR name to follow our [PR naming guidelines](https://github.com/mozilla-mobile/firefox-ios/wiki/Pull-Request-Naming-Guide)
- [ ] Wrote unit tests and/or ensured the tests suite is passing
- [x] When working on UI, I checked and implemented accessibility (minimum Dynamic Text and VoiceOver)
- [x] If needed I updated documentation / comments for complex code and public methods

### CODE REVIEW

**Comment by mattreaganmozilla** [comment]
This PR is ready for review, but adding `Do Not Merge` temporarily until all regression testing is completed.

**Comment by mattreaganmozilla** [comment]
Note: per team discussions on 1/29, it sounds like we may be pausing multi-window work temporarily in order to focus on other client improvements. Keeping this open (for now) as eventually these refactors will be needed, but we should probably wait to merge this until we can revisit multi-window in earnest.

**Review by adudenamedruby — APPROVED** [review]
LGTM.

**Comment by adudenamedruby** [comment]
@mattreaganmozilla, how will these changes affect the current operation of the app? I ask because, as things move forward, this is gonna be difficult to maintain & rebase so if we can get it in, it might be a good idea to do it sooner than later to avoid that headache.

**Comment by mattreaganmozilla** [comment]
> how will these changes affect the current operation of the app?

@adudenamedruby These refactors shouldn't impact any current behaviors on `main`; they update the associated values of the `Action`s to better support window UUIDs (this approach was decided on a while back, based on earlier team discussions), but since multi-window isn't enabled yet there is never more than one active window UUID. Additionally, all of the reducers are still updating state as before, even if the UUID is `.unavailable`. However there is a lot of code here obviously so I wanted to be sure that this was extensively regression tested before merging.

I agree with your concern about maintainability. I'm not 100% sure what the team's priorities will be over the next week or two, so I'm not sure if/when I will be able to revisit this PR. It might make sense for me to spend some time right now so that we can get this merged sooner rather than later. I will double-check with @OrlaM and Andy to just confirm if that works for our current timelines.

**Comment by adudenamedruby** [comment]
Sounds good. This is a good chunk of work and represents time invested and I'd rather not have it get lost.

**Comment by mattreaganmozilla** [comment]
@OrlaM There were considerable merge conflicts between https://github.com/mozilla-mobile/firefox-ios/pull/18340 and this PR; I've resolved them all and re-reviewed the diff and I couldn't see any problems but I am pretty bleary-eyed at this point. Any 👀 you can lend to double-check the merge would be helpful. I'm currently in the process of trying to regression test everything as well.

**Review by OrlaM — APPROVED** [review]

**Comment by OrlaM** [comment]
@MergifyIO backport release/v123

**Pull request merged**
